README.md
# Fahrscheine, bitte!

[![Build Status](https://travis-ci.org/dryewo/fahrscheine-bitte.svg?branch=master)](https://travis-ci.org/dryewo/fahrscheine-bitte)
[![codecov](https://codecov.io/gh/dryewo/fahrscheine-bitte/branch/master/graph/badge.svg)](https://codecov.io/gh/dryewo/fahrscheine-bitte)
[![Clojars Project](https://img.shields.io/clojars/v/fahrscheine-bitte.svg)](https://clojars.org/fahrscheine-bitte)

A Clojure library for verifying [OAuth2 access tokens].
For using in [Compojure] routes or in [Swagger1st] security handlers.

> "Die Fahrscheine, bitte!" is what *Fahrkartenkontrolleur* says when entering a bus. *Schwarzfahrer* are fearing this.

Access tokens are verified against [Introspection Endpoint]. In the examples in this document the address of
the endpoint is configured through `TOKENINFO_URL` environment variable.

Responses of this endpoint are by default cached for 2 minutes and only for last 100 tokens (configurable).
Creation of the cached token resolver function has to be done by the user using the provided helper function
to enable testability and follow best practices: make state explicit. Please see examples below.

## Usage

```edn
[fahrscheine-bitte "0.4.0"]
```

Examples assume the following:

```clj
(require '[fahrscheine-bitte.core :as oauth2]
         '[mount.core :as m])

(defn log-access-denied-reason [reason]
  (log/info "Access denied: %s" reason))
```

1. Example with [mount] and [Swagger1st]:

```clj
(m/defstate oauth2-s1st-security-handler
  :start (if-let [tokeninfo-url (System/getenv "TOKENINFO_URL")]
           (let [access-token-resolver-fn (oauth2/make-cached-access-token-resolver tokeninfo-url {})]
             (log/info "Checking OAuth2 access tokens against %s." tokeninfo-url)
             (oauth2/make-oauth2-s1st-security-handler access-token-resolver-fn oauth2/check-corresponding-attributes))
           (do
             (log/warn "No TOKENINFO_URL set; NOT ENFORCING SECURITY!")
             (fn [request definition requirements]
               request))))

(m/defstate handler
  :start (-> (s1st/context :yaml-cp "api.yaml")
             (s1st/discoverer)
             (s1st/mapper)
             (s1st/ring oauth2/wrap-reason-logger log-access-denied-reason)
             (s1st/protector {"oauth2" oauth2-s1st-security-handler})
             (s1st/parser)
             (s1st/executor)))
```

In this example we create a security handler that is given to `s1st/protector` to verify tokens on all endpoints that have
`oauth2` security definition in place.
Additionally, we insert a middleware `wrap-log-auth-error` that will log all rejected access attempts.

2. Example with [mount] and [Compojure]:

```clj
(m/defstate wrap-oauth2-token-verifier
  :start (if-let [tokeninfo-url (System/getenv "TOKENINFO_URL")]
           (let [access-token-resolver-fn (oauth2/make-cached-access-token-resolver tokeninfo-url {})]
             (log/info "Checking OAuth2 access tokens against %s." tokeninfo-url)
             (oauth2/make-wrap-oauth2-token-verifier access-token-resolver-fn))
           (do
             (log/warn "No TOKENINFO_URL set; NOT ENFORCING SECURITY!")
             identity)))

(defn make-handler2 []
  (-> (routes
        (GET "/hello" req {:status 200}))
      (wrap-oauth2-token-verifier)
      (oauth2/wrap-log-auth-error log-access-denied-reason)))
```

`(oauth2/make-wrap-oauth2-token-verifier access-token-resolver-fn)` returns a Ring middleware that can be used to
check access tokens against given token introspection endpoint.

### Configuring caching of tokeninfo responses

You can configure caching by passing the following parameters to `make-cached-access-token-resolver`:

* `:ttl-ms` - number of milliseconds to keep cached results. Both positive and negative results are cached.
  However, errored calls to tokeninfo are not cached. Default: 2000.
* `:max-size` - number of results to cache. This has to be limited to avoid running out of memory. Default: 100

Example:

```clj
(let [access-token-resolver-fn (oauth2/make-cached-access-token-resolver tokeninfo-url {:ttl-ms 5000 :max-size 1000})]
  (oauth2/make-wrap-oauth2-token-verifier access-token-resolver-fn))
```

### Wrapping calls to tokeninfo endpoint

For tracing or logging reasons you might want to wrap the actual HTTP GET request to TOKENINFO_URL.

This is possible by giving `:client-middleware` parameter to `oauth2/make-cached-access-token-resolver`:

```clj
;; Define the middleware
(defn wrap-client-tracing [http-get]
  (fn [url params]
    (println "Calling tokeninfo URL:" url)
    (let [res (http-get url params)]
      (println "Tokeninfo result status:" (:status res))
      res)))

;; Use the middleware when creating a token resolver
(let [access-token-resolver-fn (oauth2/make-cached-access-token-resolver tokeninfo-url {:client-middleware wrap-client-tracing})]
  (oauth2/make-wrap-oauth2-token-verifier access-token-resolver-fn))
```

`:client-middleware` is a middleware for a `clj-http.client/get`-like functions:

```clj
(ns clj-http.client)

(defn get [url params]
  ...)
```

Client middleware is a function that accepts a `clj-http.client/get`-like function —
which takes 2 parameters, `url` and `params`, and returns a response map just like `clj-http.client/get` does —
and returns another `clj-http.client/get`-like function.

You can read more in the [clj-http docs](https://github.com/dakrone/clj-http).

By default no client middleware is applied, `clj-http.client/get` is called directly.

#### Customizing HTTP client call

Client middleware can be used to adjust some parameters of `clj-http.client/get` call, for example, to set timeouts:

```clj
(defn wrap-client-timeouts [http-get]
  (fn [url params]
    ;; :conn-timeout and :socket-timeout do not overlap, giving possible duration of up to 2 seconds
    (http-get url (merge params {:conn-timeout 1000 :socket-timeout 1000}))))

;; Middlewares are combined using `comp`, first middleware is entered first
(let [access-token-resolver-fn (oauth2/make-cached-access-token-resolver tokeninfo-url
                                 {:client-middleware (comp wrap-client-tracing wrap-client-timeouts)})]
  (oauth2/make-wrap-oauth2-token-verifier access-token-resolver-fn))

``` 

**WARNING** It's not recommended to wrap the `http-get` call with `try`, as it will disrupt the logic of caching and error reporting.


## License

Copyright © 2019 Dmitrii Balakhonskii

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.

[mount]: https://github.com/tolitius/mount
[swagger1st]: https://github.com/zalando-stups/swagger1st
[Compojure]: https://github.com/weavejester/compojure
[Introspection Endpoint]: https://tools.ietf.org/html/rfc7662#section-2
[OAuth2 access tokens]: https://tools.ietf.org/html/rfc6749#section-1.4


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).


## [Unreleased]

## [0.4.0] — 2019-01-02
### Added
- Document how to configure caching.
- Document how to configure timeouts for tokeninfo calls.
### Changed
- Handlers created by `make-wrap-oauth2-token-verifier` and `make-oauth2-s1st-security-handler` never throw
  (previously threw if tokeninfo service returned 5xx or on connection errors).
### Removed
- Not using [Hystrix] because it's causing issues with tracing (call is executed on a different thread)
  and because it's not maintained anymore.

## [0.3.0] — 2018-12-18
### Added
- Client middleware support to enable wrapping calls to tokeninfo URL.
### Changed
- Use Clojure 1.10
- Bumped library versions.

## [0.2.2] — 2018-09-03
### Added
- _CHANGELOG.md_ created.
- Automated version bump in _README.md_.
### Changed
- Bumped library versions.

## 0.2.1 — 2018-06-04
Released without _CHANGELOG.md_.


[0.2.2]: https://github.com/dryewo/fahrscheine-bitte/compare/0.2.1...0.2.2
[0.3.0]: https://github.com/dryewo/fahrscheine-bitte/compare/0.2.2...0.3.0

[Hystrix]: https://github.com/Netflix/Hystrix#hystrix-status
[0.4.0]: https://github.com/dryewo/fahrscheine-bitte/compare/0.3.0...0.4.0
[Unreleased]: https://github.com/dryewo/fahrscheine-bitte/compare/0.4.0...HEAD


test/fahrscheine_bitte/core_test.clj
(ns fahrscheine-bitte.core-test
  (:require [clojure.test :refer :all]
            [fahrscheine-bitte.core :refer :all :as v]
            [juxt.iota :refer [given]]
            [clj-http.client :as http]))


(defn resolve-token-valid [token]
  {"scope"        ["uid"]
   "access_token" token
   "uid"          "mjackson"})


(defn resolve-token-invalid [token]
  nil)


(defn resolve-token-error [token]
  (throw (IllegalStateException. (str "tokeninfo endpoint returned status code: " 503))))


(defn wrap-client-tracing [http-get]
  (fn [url params]
    (println "Calling tokeninfo URL:" url)
    (let [res (http-get url (assoc params :foo "bar"))]
      (println "Tokeninfo result status:" (:status res))
      res)))


(def handler1 (make-oauth2-s1st-security-handler resolve-token-valid check-consented-scopes))
(def handler2 (make-oauth2-s1st-security-handler resolve-token-invalid (constantly true)))
(def handler4 (make-oauth2-s1st-security-handler resolve-token-valid check-corresponding-attributes))


(deftest s1st-handler

  (testing "When token is valid"
    (given (handler1 {} nil [])
      :status := 401
      ::v/reason-code :token-missing)
    (given (handler1 {:headers {"authorization" "Bearer footoken"}} nil ["foo.read"])
      :status := 403
      ::v/reason-code := :scopes)
    (given (handler1 {:headers {"authorization" "Bearer footoken"}} nil ["uid"])
      :tokeninfo :> {"scope" ["uid"] "access_token" "footoken"}
      :status := nil
      ::v/reason-code := nil)
    (given (handler4 {:headers {"authorization" "Bearer footoken"}} nil ["foo.read"])
      :status := 403
      ::v/reason-code := :scopes)
    (given (handler4 {:headers {"authorization" "Bearer footoken"}} nil ["uid"])
      :tokeninfo :> {"uid" "mjackson" "access_token" "footoken"}
      :status := nil
      ::v/reason-code := nil))

  (testing "When token is invalid"
    (given (handler2 {} nil [])
      :status := 401
      ::v/reason-code :token-missing)
    (given (handler2 {:headers {"authorization" "Bearer footoken"}} nil [])
      :status := 401
      ::v/reason-code :token-invalid))

  (testing "integration"
    (testing "when everything is ok"
      (let [handler         (make-oauth2-s1st-security-handler (make-cached-access-token-resolver "tokeninfo-url" {})
                                                               check-consented-scopes)
            http-get-called (atom [])]
        (with-redefs [http/get (fn [url {:keys [oauth-token]}]
                                 (swap! http-get-called conj [url oauth-token])
                                 {:status 200 :body {"access_token" oauth-token "uid" "mjackson" "scope" ["uid"]}})]
          (testing "http/get should be called"
            (given (handler {:headers {"authorization" "Bearer footoken"}} nil ["uid"])
              ;; :status being nil means we still have a request, not a response - normal operation
              :status := nil
              ;; :tokeninfo key is injected into the request
              :tokeninfo :> {"scope" ["uid"], "access_token" "footoken"})
            (is (= [["tokeninfo-url" "footoken"]] @http-get-called)))
          (testing "Call again, cached response should be used"
            (handler {:headers {"authorization" "Bearer footoken"}} nil ["uid"])
            (is (= 1 (count @http-get-called)))))))

    (testing "when http/get returns 5xx"
      (let [handler (make-oauth2-s1st-security-handler (make-cached-access-token-resolver "tokeninfo-url" {})
                                                       check-consented-scopes)]
        (with-redefs [http/get (fn [_ _] {:status 503})]
          (given (handler {:headers {"authorization" "Bearer footoken"}} nil [])
            :status := 504
            :body := "access-token-resolver-fn call threw an exception: java.lang.IllegalStateException: tokeninfo endpoint returned status code: 503"))))

    (testing "when http/get throws an exception"
      (let [handler (make-oauth2-s1st-security-handler (make-cached-access-token-resolver "tokeninfo-url" {})
                                                       check-consented-scopes)]
        (with-redefs [http/get (fn [_ _] (throw (Exception. "EXPLODED!")))]
          (given (handler {:headers {"authorization" "Bearer footoken"}} nil [])
            :status := 504
            :body := "access-token-resolver-fn call threw an exception: java.lang.Exception: EXPLODED!"))))

    (testing "wrapped http/get call"
      (let [handler (make-oauth2-s1st-security-handler (make-cached-access-token-resolver "tokeninfo-url"
                                                                                          {:client-middleware wrap-client-tracing})
                                                       check-consented-scopes)]
        (with-redefs [http/get (fn [url {:keys [oauth-token foo]}]
                                 {:status 200 :body {"access_token" oauth-token "uid" "mjackson" "scope" ["uid"] "foo" foo}})]
          (testing "http/get is called with a middleware"
            (given (handler {:headers {"authorization" "Bearer footoken"}} nil ["uid"])
              ;; :status being nil means we still have a request, not a response - normal operation
              :status := nil
              ;; Injected parameter is passed back, this proves that our middleware was used
              :tokeninfo :> {"foo" "bar"})))))))


(defn ring-handler [request]
  {:status 200 :body {:incoming-tokeninfo (:tokeninfo request)}})


(def wrapped-ring-handler1 (-> ring-handler ((make-wrap-oauth2-token-verifier resolve-token-valid))))
(def wrapped-ring-handler2 (-> ring-handler ((make-wrap-oauth2-token-verifier resolve-token-invalid))))
(def wrapped-ring-handler3 (-> ring-handler ((make-wrap-oauth2-token-verifier resolve-token-error))))


(defn error-logger [message]
  (println "Access denied:" message))


(deftest ring-middleware

  (given (wrapped-ring-handler1 {})
    :status := 401
    ::v/reason-code :token-missing)
  (given (wrapped-ring-handler1 {:headers {"authorization" "Bearer footoken"}})
    [:body :incoming-tokeninfo] :> {"scope" ["uid"] "access_token" "footoken"}
    :status := 200
    ::v/reason-code := nil)

  (testing "When token is invalid"
    (given (wrapped-ring-handler2 {})
      :status := 401
      ::v/reason-code :token-missing)
    (given (handler2 {:headers {"authorization" "Bearer footoken"}} nil [])
      :status := 401
      ::v/reason-code :token-invalid))

  (testing "When tokeninfo returns 5xx"
    (given (wrapped-ring-handler3 {:headers {"authorization" "Bearer footoken"}})
      :status := 504))

  (testing "integration"
    (let [messages  (atom [])
          logger-fn (fn [msg]
                      (swap! messages conj msg))]
      (given ((wrap-log-auth-error wrapped-ring-handler1 logger-fn) {})
        :status := 401)
      (given ((wrap-log-auth-error wrapped-ring-handler2 logger-fn) {:headers {"authorization" "Bearer footoken"}})
        :status := 401)
      (given ((wrap-log-auth-error wrapped-ring-handler3 logger-fn) {:headers {"authorization" "Bearer footoken"}})
        :status := 504)
      (is (= ["no access token given"
              "invalid access token"
              "access-token-resolver-fn call threw an exception: java.lang.IllegalStateException: tokeninfo endpoint returned status code: 503"]
             @messages)))))


project.clj
(defproject fahrscheine-bitte "0.4.0"
  :description "Library for verifying OAuth2 access tokens"
  :url "https://github.com/dryewo/fahrscheine-bitte"
  :license {:name "Eclipse Public License"
            :url  "http://www.eclipse.org/legal/epl-v10.html"}
  :dependencies [[clj-http "3.9.1"]
                 [org.clojure/core.memoize "0.7.1"]]
  :plugins [[lein-cloverage "1.0.13"]
            [lein-ancient "0.6.15"]
            [lein-shell "0.5.0"]
            [lein-changelog "0.3.2"]]
  :profiles {:dev {:dependencies [[org.clojure/clojure "1.10.0"]
                                  [juxt/iota "0.2.3"]]}}
  :deploy-repositories [["releases" :clojars]]
  :aliases {"update-readme-version" ["shell" "sed" "-i" "s/\\\\[fahrscheine-bitte \"[0-9.]*\"\\\\]/[fahrscheine-bitte \"${:version}\"]/" "README.md"]}
  :release-tasks [["shell" "git" "diff" "--exit-code"]
                  ["change" "version" "leiningen.release/bump-version"]
                  ["change" "version" "leiningen.release/bump-version" "release"]
                  ["changelog" "release"]
                  ["update-readme-version"]
                  ["vcs" "commit"]
                  ["vcs" "tag"]
                  ["deploy"]
                  ["vcs" "push"]])


src/fahrscheine_bitte/core.clj
(ns fahrscheine-bitte.core
  (:require [clj-http.client :as http]
            [clojure.core.memoize :as memo]
            [clojure.core.cache :as cache]))


(defn extract-access-token
  "Extracts the Bearer token from the Authorization header.
   Returns either nil or the token string."
  [request]
  (if-let [authorization (get-in request [:headers "authorization"])]
    (when (.startsWith authorization "Bearer ")
      (.substring authorization (count "Bearer ")))))


(defn check-consented-scopes
  "Checks if every scope is mentioned in the 'scope' attribute of the token info:
   {\"scope\" [\"read\" \"write\"]}"
  [tokeninfo required-scopes]
  (let [consented-scopes (set (get tokeninfo "scope"))]
    (every? #(contains? consented-scopes %) required-scopes)))


(defn check-corresponding-attributes
  "Checks if every scope has a truthy attribute in the token info of the same name:
   {\"read\": true, \"write\": true}"
  [tokeninfo required-scopes]
  (let [scope-as-attribute-true? (fn [scope]
                                   (get tokeninfo scope))]
    (every? scope-as-attribute-true? required-scopes)))


(defn fetch-tokeninfo
  "Returns a response with status 2xx or 4xx,
   throws an exception on 5xx or when http/get throws.
   Returned results are cacheable, thrown ones are not."
  [tokeninfo-url access-token client-middleware]
  ;; http/get with {:throw-exceptions false} throws only on connection errors or timeouts
  (let [response ((client-middleware http/get) tokeninfo-url {:oauth-token      access-token
                                                              :as               :json-string-keys
                                                              :throw-exceptions false})]
    ;; If returned 5xx, throw an exception, so that the response is not cached
    ;; Caching 4xx is ok
    (if (http/server-error? response)
      (throw (IllegalStateException. (str "tokeninfo endpoint returned status code: " (:status response))))
      response)))


(defn resolve-access-token
  "Checks with a tokeninfo endpoint for the token's validity.
   Returns the session information if valid (2xx), otherwise (4xx) returns nil.
   Throws if fetch-tokeninfo throws (which it does on 5xx or on connection errors/timeouts)."
  [tokeninfo-url access-token client-middleware]
  (let [response (fetch-tokeninfo tokeninfo-url access-token client-middleware)]
    (when (http/success? response)
      (:body response))))


(defn make-cached-access-token-resolver
  "Returns a function that throws or returns whatever resolve-access-token throws or returns.
   Only returned results are cached."
  [tokeninfo-url {:keys [ttl-ms
                         max-size
                         client-middleware]
                  :or   {ttl-ms            120000
                         max-size          100
                         client-middleware identity}}]
  ;; Caches only the results returned normally:
  ;; 2xx:                                 response body is cached
  ;; 4xx:                                 nil is cached
  ;; exception from resolve-access-token: nothing is cached
  (memo/fifo #(resolve-access-token tokeninfo-url % client-middleware)
             (cache/ttl-cache-factory {} :ttl ttl-ms) :fifo/threshold max-size))


(defn make-wrap-oauth2-token-verifier
  "Returns a swagger1st security handler that checks OAuth 2.0 tokens.
   * access-token-resolver-fn takes a token and returns tokeninfo: https://tools.ietf.org/html/rfc7662#section-2.2
   access-token-resolver-fn is presumably created by make-cached-access-token-resolver."
  [access-token-resolver-fn]
  (fn [next-handler]
    ;; Never throws, sometimes calls next-handler, sometimes returns response with 401 or 504
    ;; token missing: returns 401
    ;; then, depending on tokeninfo call result:
    ;;   2xx (can come from cache)                           - injects returned tokeninfo into the request,
    ;;                                                         calls next handler
    ;;   4xx (can come from cache)                           - returns a 401 response
    ;;   thrown (because of 5xx or connection error/timeout) - returns a 504 response
    (fn [request]
      (let [access-token (extract-access-token request)]
        (if-not access-token
          {:status 401 ::reason-code :token-missing}
          (let [tokeninfo (try
                            (access-token-resolver-fn access-token)
                            (catch Exception e
                              e))]
            (cond
              (nil? tokeninfo)
              {:status 401 ::reason-code :token-invalid}

              (instance? Exception tokeninfo)
              (let [message (str "access-token-resolver-fn call threw an exception: " (str tokeninfo))]
                {:status 504 :body message ::reason-code message})

              :else
              (next-handler (assoc request :tokeninfo tokeninfo)))))))))


(defn make-oauth2-s1st-security-handler
  "Returns a swagger1st security handler that checks OAuth 2.0 tokens.
   * access-token-resolver-fn takes a token and returns tokeninfo: https://tools.ietf.org/html/rfc7662#section-2.2
   * scope-checker-fn takes tokeninfo and requirements and returns true if scopes in the tokeninfo match the requirements
   access-token-resolver-fn is presumably created by make-cached-access-token-resolver."
  [access-token-resolver-fn scope-checker-fn]
  ;; Never throws.
  ;; Returns the same request with :tokeninfo injected if everything is ok
  ;; or a response map with :status key if something went wrong
  (fn [request _ requirements]
    ;; Very sophisticated way to reuse logic from make-wrap-oauth2-token-verifier
    ;; Avoiding code duplication at all costs
    (let [middleware-that-calls-tokeninfo                  (make-wrap-oauth2-token-verifier access-token-resolver-fn)
          handler-that-checks-scopes-in-injected-tokeninfo (middleware-that-calls-tokeninfo
                                                             ;; This handler expects :tokeninfo key present in request
                                                             ;; Just checks if scopes are satisfied
                                                             ;; Returns original request if everything is ok,
                                                             ;;  response with 403 otherwise
                                                             (fn [req]
                                                               (if-not (scope-checker-fn (:tokeninfo req) requirements)
                                                                 {:status 403 ::reason-code :scopes}
                                                                 req)))]
      (handler-that-checks-scopes-in-injected-tokeninfo request))))


(defn explain-reason-code
  "If code is one of the supported codes, returns the explanation.
   If code is a string, returns it as is."
  [code]
  (if (keyword? code)
    (get {:scopes        "scopes not granted"
          :token-invalid "invalid access token"
          :token-missing "no access token given"}
         code)
    code))


(defn wrap-log-auth-error [next-handler logger-fn]
  (fn [request]
    (let [response (next-handler request)]
      (when-let [reason (explain-reason-code (::reason-code response))]
        (logger-fn reason))
      response)))
